src/util/file_bytes_stream.rs

```diff
@@ -1,25 +1,126 @@
-        FileBytesStream { file, buf }
+        FileBytesStream { file, buf, range_bytes: None }
+    }
+
+    /// Set ranges if the client requested it.  Boundary must be the empty string if there is only one
+    /// range.  It is invalid to pass a list of zero ranges.
+    pub fn set_ranges(&mut self, ranges: Vec<HttpRange>, boundary: String, file_length: u64) -> &mut Self {
+        assert_eq!(ranges.len() > 1, boundary.len() > 0);
+        let is_first_boundary = ranges.len() > 1;
+        self.range_bytes = Some(RangeBytes {
+            seek_completed: true,
+            range_remaining: 0,
+            file_length,
+            ranges: ranges.into_iter(),
+            boundary,
+            is_first_boundary,
+        });
+        self
+    }
+}
+
+impl RangeBytes {
+
+    pub fn handle(&mut self, file: &mut File, buf: &mut [MaybeUninit<u8>], cx: &mut Context) -> Poll<Option<Result<Bytes, IoError>>> {
+        loop {
+            if !self.seek_completed {
+                match Pin::new(&mut *file).poll_complete(cx) {
+                    Poll::Ready(Ok(..)) => (),
+                    Poll::Ready(Err(e)) => return Poll::Ready(Some(Err(e))),
+                    Poll::Pending => return Poll::Pending,
+                }
+            }
+
+            // we've finished the previous range (or we're just starting), we need to seek
+            // and repopulate the remaining field if we're not completely finished.
+            if self.range_remaining == 0 {
+                let range = match self.ranges.next() {
+                    Some(r) => r,
+                    None => return Poll::Ready(None),
+                };
+                if let Err(e) = Pin::new(&mut *file).start_seek(SeekFrom::Start(range.start)) {
+                    return Poll::Ready(Some(Err(e)));
+                }
+                self.seek_completed = false;
+                self.range_remaining = range.length;
+                if self.boundary.len() > 0 {
+                    let mut mp_header = Vec::new();
+                    if self.is_first_boundary {
+                        self.is_first_boundary = false;
+                    } else {
+                        write!(&mut mp_header, "\r\n").unwrap();
+                    }
+                    write!(&mut mp_header, "--{}\r\n", self.boundary).unwrap();
+                    write!(
+                        &mut mp_header,
+                        "Content-Range: bytes {}-{}/{}\r\n",
+                        range.start,
+                        range.start + range.length - 1,
+                        self.file_length,
+                    )
+                    .unwrap();
+                    // write!(&mut mp_header, "Content-Type: {}\r\n", file_mime_type).unwrap();
+                    write!(&mut mp_header, "\r\n").unwrap();
+
+                    return Poll::Ready(Some(Ok(mp_header.into())));
+                }
+
+                // back to the top of the loop so we can poll on the seek's completion
```

Comment:
might be worth to de-loop this since we're guaranteed to hit the file reading stage after the completed seek.

Reply:
Fixed